rust/protocol/src/crypto.rs

```diff
@@ -1,156 +1,244 @@
 //
-// Copyright 2020 Signal Messenger, LLC.
+// Copyright 2020-2022 Signal Messenger, LLC.
 // SPDX-License-Identifier: AGPL-3.0-only
 //
 
-use std::convert::TryInto;
-use std::result::Result;
+//! Internal convenience wrappers for symmetric cryptographic primitives.
 
 use aes::cipher::{NewCipher, StreamCipher};
 use aes::{Aes256, Aes256Ctr};
+use arrayref::array_ref;
 use block_modes::block_padding::Pkcs7;
 use block_modes::{BlockMode, Cbc};
+use displaydoc::Display;
 use hmac::{Hmac, Mac, NewMac};
 use sha2::Sha256;
 use subtle::ConstantTimeEq;
+use thiserror::Error;
 
-#[derive(Debug)]
-pub(crate) enum EncryptionError {
-    /// The key or IV is the wrong length.
-    BadKeyOrIv,
-}
+/// bad ciphertext: {0}
+///
+/// Either the input is malformed, or the MAC doesn't match on decryption.
+///
+/// These cases should not be distinguished; message corruption can cause either problem.
+#[derive(Debug, Display, Error)]
+#[ignore_extra_doc_attributes]
+pub struct DecryptionError(pub &'static str);
 
-#[derive(Debug)]
-pub(crate) enum DecryptionError {
-    /// The key or IV is the wrong length.
-    BadKeyOrIv,
-    /// Either the input is malformed, or the MAC doesn't match on decryption.
-    ///
-    /// These cases should not be distinguished; message corruption can cause either problem.
-    BadCiphertext(&'static str),
-}
+// TODO: Could be nice to have a type-safe library for manipulating units of bytes safely.
+const BITS_PER_BYTE: usize = std::mem::size_of::<u8>() * 8;
 
-fn aes_256_ctr_encrypt(ptext: &[u8], key: &[u8]) -> Result<Vec<u8>, EncryptionError> {
-    let key: [u8; 32] = key.try_into().map_err(|_| EncryptionError::BadKeyOrIv)?;
+/// The length of the key we use for AES encryption in this crate.
+pub const AES_256_KEY_SIZE: usize = 256 / BITS_PER_BYTE;
 
-    let zero_nonce = [0u8; 16];
-    let mut cipher = Aes256Ctr::new(key[..].into(), zero_nonce[..].into());
+/// The size of the generated nonce we use for AES encryption in this crate.
+pub const AES_NONCE_SIZE: usize = 128 / BITS_PER_BYTE;
 
-    let mut ctext = ptext.to_vec();
-    cipher.apply_keystream(&mut ctext);
-    Ok(ctext)
-}
+/// Use AES-256 in CTR mode.
+pub mod ctr {
```

Comment:
I don't know if I really see the point of this. Yes, there's a little more organization, but "CTR" and "CBR" don't imply "AES-256" by themselves, and they're all just convenience wrappers for the session implementation in any case.